pkg/cmd/util/tokencmd/request_token.go

```diff
@@ -1,121 +1,148 @@
 package tokencmd
 
 import (
-	"bytes"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"net/http"
 	"net/url"
 
-	"github.com/RangelReale/osincli"
-	"github.com/golang/glog"
-
+	"k8s.io/kubernetes/pkg/api"
+	apierrs "k8s.io/kubernetes/pkg/api/errors"
 	kclient "k8s.io/kubernetes/pkg/client"
-
-	"github.com/openshift/origin/pkg/client"
-	"github.com/openshift/origin/pkg/oauth/server/osinserver"
+	"k8s.io/kubernetes/pkg/util"
 )
 
+// CSRFTokenHeader is a marker header that indicates we are not a browser that got tricked into requesting basic auth
+// Corresponds to the header expected by basic-auth challenging authenticators
+const CSRFTokenHeader = "X-CSRF-Token"
+
 // RequestToken uses the cmd arguments to locate an openshift oauth server and attempts to authenticate
 // it returns the access token if it gets one.  An error if it does not
 func RequestToken(clientCfg *kclient.Config, reader io.Reader, defaultUsername string, defaultPassword string) (string, error) {
-	tokenGetter := &tokenGetterInfo{}
-
-	osClient, err := client.New(clientCfg)
-	if err != nil {
-		return "", err
+	challengeHandler := &BasicChallengeHandler{
+		Host:     clientCfg.Host,
+		Reader:   reader,
+		Username: defaultUsername,
+		Password: defaultPassword,
 	}
 
-	// get the transport, so that we can use it to build our own client that wraps it
-	// our client understands certain challenges and can respond to them
-	clientTransport, err := kclient.TransportFor(clientCfg)
+	rt, err := kclient.TransportFor(clientCfg)
 	if err != nil {
 		return "", err
 	}
 
-	httpClient := &http.Client{
-		Transport:     clientTransport,
-		CheckRedirect: tokenGetter.checkRedirect,
-	}
-
-	osClient.Client = &challengingClient{httpClient, reader, defaultUsername, defaultPassword}
+	// requestURL holds the current URL to make requests to. This can change if the server responds with a redirect
+	requestURL := clientCfg.Host + "/oauth/authorize?response_type=token&client_id=openshift-challenging-client"
+	// requestHeaders holds additional headers to add to the request. This can be changed by challengeHandlers
+	requestHeaders := http.Header{}
+	// requestedURLs holds the URLs we have requested, to prevent redirect loops. Gets reset when a challenge is handled.
+	requestedURLs := util.NewStringSet()
+
+	for {
```

Comment:
they're not retries... looping means one of the following happened:
* you got redirected to new URL (we prevent cycles)
* you got challenged and a challenge handler indicated you should retry (all handlers should limit number of retries... the basic one only attempts a single time)